File: apps/client/lib/club_accounting/contracts.dart
import 'dart:convert';
import 'dart:math';

String clubRequestKey() {
  final random = Random.secure();
  return 'club-${List.generate(16, (_) => random.nextInt(256).toRadixString(16).padLeft(2, '0')).join()}';
}
String clubDate(DateTime date) => '${date.year.toString().padLeft(4, '0')}-${date.month.toString().padLeft(2, '0')}-${date.day.toString().padLeft(2, '0')}';
String clubToday() => clubDate(DateTime.now().toUtc().add(const Duration(hours: 3)));
String clubLabel(String value) => value.replaceAll('_', ' ').replaceAll('minor', '(minor units)').replaceAll('micro', '(micro-units)');
List<Map<String, dynamic>> clubRows(dynamic value) => value is List ? value.whereType<Map>().map((row) => row.cast<String, dynamic>()).toList() : [];
int clubInteger(dynamic value, [int minimum = 0]) {
  final text = (value ?? '').toString().trim();
  if (!RegExp(r'^[0-9]+$').hasMatch(text)) throw const FormatException('Enter a whole number without commas or decimals.');
  final number = int.tryParse(text);
  if (number == null || number < minimum || number > 9007199254740991) throw const FormatException('The number is outside the supported range.');
  return number;
}
String clubIsoDate(String text) {
  final parsed = DateTime.tryParse(text);
  if (!RegExp(r'^\d{4}-\d{2}-\d{2}$').hasMatch(text) || parsed == null || clubDate(parsed) != text) {
    throw const FormatException('Choose a real calendar date.');
  }
  return text;
}
Map<String, dynamic> clubEmptyFields(List<Map<String, dynamic>> fields) => {
  for (final field in fields) field['key'] as String: field['type'] == 'array' ? <dynamic>[] : '',
};
Map<String, dynamic> clubValidate(List<Map<String, dynamic>> fields, Map<String, dynamic> raw) {
  final result = <String, dynamic>{};
  for (final field in fields) {
    final key = field['key'] as String;
    final value = raw[key];
    final label = field['label']?.toString() ?? key;
    if (value == null || value == '') {
      if (field['required'] == true) throw FormatException('$label is required.');
      continue;
    }
    switch (field['type']) {
      case 'array':
        if (value is! List || value.length > 1000) throw FormatException('$label has invalid or too many rows.');
        result[key] = value.map((row) {
          if (field['items_type'] == 'integer') return clubInteger(row, 1);
          if (row is! Map) throw FormatException('Invalid $label row.');
          return clubValidate(clubRows(field['items']), row.cast<String, dynamic>());
        }).toList();
        break;
      case 'integer':
        final number = clubInteger(value, field['minimum'] is int ? field['minimum'] as int : 0);
        if (field['maximum'] is num && number > (field['maximum'] as num)) throw FormatException('$label is too large.');
        result[key] = number;
        break;
      case 'string': case 'date': case 'month':
        if (value is! String) throw FormatException('$label must be text.');
        final text = value.trim();
        if (text.length > (field['maxLength'] is int ? field['maxLength'] as int : 1000)) throw FormatException('$label is too long.');
        if (field['enum'] is List && !(field['enum'] as List).contains(text)) throw FormatException('Choose a supported $label.');
        if (field['type'] == 'date') clubIsoDate(text);
        if (field['type'] == 'month') clubIsoDate('$text-01');
        result[key] = text;
        break;
      default: throw FormatException('Unsupported form field: $label. Refresh the application contract.');
    }
  }
  for (final key in raw.keys) {
    if (!fields.any((field) => field['key'] == key)) throw FormatException('Unsupported field: $key');
  }
  return result;
}
Map<String, dynamic> clubCopy(Map<String, dynamic> value) => (jsonDecode(jsonEncode(value)) as Map).cast<String, dynamic>();

File: apps/web/src/lib/club-accounting/contracts.ts
export type Json = null | boolean | number | string | Json[] | { [key: string]: Json };
export type RecordValue = { [key: string]: Json };
export type Field = {
  key: string; label: string; type: 'integer' | 'string' | 'date' | 'month' | 'array';
  required: boolean; source?: string; enum?: string[]; items?: Field[]; items_type?: string;
  minimum?: number; maximum?: number; maxLength?: number;
};
export type Operation = { type: string; title: string; effect: string; fields: Field[] };
export type Instruction = {
  id: number; reference: string; type: string; business_date: string; status: string;
  maker_id: number; checker_id?: number; payload_hash: string; payload: RecordValue; result?: RecordValue;
};
export type Book = {
  id: number; public_id: string; currency: string; ownership_model: string; status: string;
  cutover_date: string; closed_through?: string; my_position?: RecordValue;
};
export type ClubAction = 'schema' | 'books' | 'create-book' | 'catalogue' | 'instructions' |
  'submit' | 'instruction' | 'preview' | 'approve' | 'reject' | 'cancel' |
  'report' | 'journals' | 'integrity' | 'statements' | 'issue-statement' | 'statement' | 'profile';
export type ClubResult = { ok: true; data: RecordValue } | { ok: false; message: string; status?: number };

export function positiveId(value: unknown): number {
  if (typeof value !== 'number' || !Number.isSafeInteger(value) || value <= 0) throw new Error('Choose a valid record.');
  return value;
}
export function integerInput(value: unknown, minimum = 0): number {
  const text = String(value ?? '').trim();
  if (!/^[0-9]+$/.test(text)) throw new Error('Enter a whole number, without commas or decimals.');
  const number = Number(text);
  if (!Number.isSafeInteger(number) || number < minimum) throw new Error('The number is outside the supported range.');
  return number;
}
export function isoDate(value: string): string {
  if (!/^\d{4}-\d{2}-\d{2}$/.test(value)) throw new Error('Choose a calendar date.');
  const date = new Date(value + 'T00:00:00.000Z');
  if (!Number.isFinite(date.getTime()) || date.toISOString().slice(0,10) !== value) throw new Error('Choose a real calendar date.');
  return value;
}
export function validateFields(fields: Field[], raw: RecordValue): RecordValue {
  const result: RecordValue = {};
  for (const field of fields) {
    const value = raw[field.key];
    if (value === undefined || value === null || value === '') {
      if (field.required) throw new Error(field.label + ' is required.');
      continue;
    }
    if (field.type === 'array') {
      if (!Array.isArray(value) || value.length > 1000) throw new Error(field.label + ' has too many or invalid rows.');
      result[field.key] = value.map(row => {
        if (field.items_type === 'integer') return integerInput(row,1);
        if (row === null || Array.isArray(row) || typeof row !== 'object') throw new Error('Invalid ' + field.label + ' row.');
        return validateFields(field.items ?? [], row);
      });
    } else if (field.type === 'integer') {
      const number = integerInput(value,field.minimum ?? 0);
      if (field.maximum !== undefined && number > field.maximum) throw new Error(field.label + ' is too large.');
      result[field.key] = number;
    } else {
      if (typeof value !== 'string') throw new Error(field.label + ' must be text.');
      const text = value.trim();
      if (text.length > (field.maxLength ?? 1000)) throw new Error(field.label + ' is too long.');
      if (field.enum && !field.enum.includes(text)) throw new Error('Choose a supported ' + field.label + '.');
      if (field.type === 'date') isoDate(text);
      if (field.type === 'month') isoDate(text + '-01');
      result[field.key] = text;
    }
  }
  for (const key of Object.keys(raw)) if (!fields.some(field => field.key === key)) throw new Error('Unsupported field: ' + key);
  return result;
}
export function sameIntent(left: RecordValue, right: RecordValue): boolean {
  const canonical = (value: Json): string => {
    if (Array.isArray(value)) return '[' + value.map(canonical).join(',') + ']';
    if (value !== null && typeof value === 'object') return '{' + Object.keys(value).sort().map(key => JSON.stringify(key)+':'+canonical(value[key])).join(',') + '}';
    return JSON.stringify(value);
  };
  return canonical(left) === canonical(right);
}
export function humanLabel(text: string): string {
  return text.replace(/_/g,' ').replace(/\bminor\b/g,'(minor units)').replace(/\bmicro\b/g,'(micro-units)');
}
